tools/autograd/gen_variable_type.py

```diff
@@ -455,24 +458,77 @@ def declare_returned_variables():
         return '\n'.join(names)
 
     def wrap_output(call):
+        # Returns a 2-tuple `(wrapped_call, extra_wrapping_stmts)`, where
+        # `wrapped_call` is to drop-in replace `call`, and
+        # `extra_wrapping_stmts` is a list of extra statements to run after
+        # `call`.
         if 'Tensor' not in declaration['return_type']:
-            return call
-        elif is_view:
-            return 'as_view(self, {})'.format(call)
+            return call, []
+        elif view_info is not None:
+            # See NOTE [ Autograd View Variables ] in variable.h for details.
+            differentiable_output_vars = {r['name'] for r in differentiable_outputs}
+            tensor_output_vars = {r['name'] for r in returns if 'Tensor' in r['type']}
+            if not isinstance(view_info, dict):
+                if len(differentiable_output_vars) == len(tensor_output_vars):
+                    # all outputs are differentiable
+                    return 'as_view({}, {}, true)'.format(view_info, call), []
+                elif len(differentiable_output_vars) == 0:
+                    # no output is differentiable
+                    return 'as_view({}, {}, false)'.format(view_info, call), []
+                else:
+                    # some of the outputs are differentiable
+                    # need to expand to dict mode, i.e., one entry per output
+                    base_name = view_info
+                    view_info_dict = {}
+                    for i, return_info in enumerate(returns):
+                        if 'Tensor' in return_info['type']:
+                            view_info_dict[i] = base_name
+            else:
+                view_info_dict = view_info
+
+            def wrap_view_single(output_var, base_var):
+                fmt = '{output_var} = as_view({base_var}, {output_var}, {is_differentiable});'
+                if output_var in differentiable_output_vars:
+                    # If `GradMode::is_enabled()` is False, this is a
+                    # non-differentiable view. Gradients should not flow through.
+                    is_differentiable = 'true'
```

Comment:
@albanD Here `is_differentiable` is always `true` regardless of `GradMode` (sorry about the comments, I forgot to update them). While in #12502 this is `'as_view(self, {}, GradMode::is_enabled())'.format(call)`.

Reply:
Alright. Thanks for the info !